### PR TITLE
Avoid circular reference with `defer.inlineCallbacks`

### LINKED ISSUE #11885: `_inlineCallbacks` machinery has a circular reference
The `_inlineCallbacks` machinery currently causes a circular reference to be made, which stops Python from freeing the deferreds etc immediately. 

This is caused by usages of inline functions. The fix is to move those functions out into standalone ones. (PR incoming).

Similar to https://github.com/twisted/twisted/issues/10310.

### PR DESCRIPTION
**Pull request by erikjohnston**
## Scope and purpose

Fixes #11885

~~Requires #11884 to work, and so is based on that (otherwise the test fails).~~

This PR does make the marginally code style worse, in that we now have a bunch of free functions that really should be bundled together. However, given how frequently `_inlineCallbacks` is used (and not just for `@inlineCallbacks`) I think this trade-off is worth it to for the memory performance.

### CODE REVIEW

**Comment by erikjohnston** [comment]
I've merged in https://github.com/twisted/twisted/issues/11885, so this should now be ready to be properly looked at.

I'm not sure why the code coverage is failing, I tried to add a test to hit the partial line, but I don't fully grok the machinery going on in there.

I also cheekily tried to make `twisted.test.test_twistd.UnixApplicationRunnerStartApplicationTests.test_setUidSameAsCurrentUid` test less flakey (c.f. [this test run](https://github.com/twisted/twisted/actions/runs/5435987737/jobs/9885527570))

**Comment by erikjohnston** [comment]
> I'm not sure why the code coverage is failing, I tried to add a test to hit the partial line, but I don't fully grok the machinery going on in there.

Ah, I managed to work that out!

File: src/twisted/test/test_twistd.py [hunk 1]
```diff
@@ -1014,6 +1014,10 @@ def test_setUidSameAsCurrentUid(self):
         If the specified UID is the same as the current UID of the process,
         then a warning is displayed.
         """
+
+        # Flush any existing warnings.
+        self.flushWarnings()
```
Comment (adiroiban) [inline src/twisted/test/test_twistd.py]: I don't this that this is a good idea.

I guess that the warning is generated by another test and that test is not checking/clearing the warnings, and it breaks here.

I think it's best to see which tests generated the warning and fix that one here.

But there is one thing that can be done here to help troubleshoot this.

Instead of checking the length first, we can check the warning content first and this might give an idea of which test is generating the other warning.

```python
        warningsShown = self.flushWarnings()
        expectedWarning = (
            "tried to drop privileges and setuid {} but uid is already {}; "
            "should we be root? Continuing.".format(currentUid, currentUid)
        )
        self.assertEqual(expectedWarning, warningsShown[0]["message"])
        # Make sure we don't have more warnings.
        self.assertEqual(1, len(warningsShown))
```
The best option is to have a dedicated warning assertion, which will also show the traceback of the warning, in the case of a warning.

I think that fixing flaky test is very important... but just ignoring the error will not help... as next time it will be another test that will fail

File: src/twisted/test/test_defer.py [hunk 2]
```diff
@@ -1681,6 +1683,37 @@ def test_fromCoroutineRequiresCoroutine(self) -> None:
...
 
+    def test_inlineCallbacksCancelCaptured(self) -> None:
+        """
+        Cancelling an `inlineCallbacks` correctly handles the function catching
+        the `CancelledError`.
+        """
+        cancellerMock1 = Mock()
+        cancellerMock2 = Mock()
+        d1: Deferred[Any] = Deferred(cancellerMock1)
+        d2: Deferred[Any] = Deferred(cancellerMock2)
```
Comment (adiroiban) [inline src/twisted/test/test_defer.py]: I am not a fan of Mock, expecilly one without a specification 

Maybe something like this

```suggestion
        canceller1Calls = []
        canceller2Calls = []
        d1: Deferred[Any] = Deferred(lambda d: canceller1Calls.append(d))
        d2: Deferred[Any] = Deferred(lambda d: canceller2Calls.append(d))
```


My biggest issue with Mock assertion is that for example if you call

```diff
-cancellerMock1.assert_called_once()
+cancellerMock1.asert_called_once()
```

it will still pass

File: src/twisted/test/test_twistd.py (hunk 1 above)
Comment (erikjohnston) [reply src/twisted/test/test_twistd.py]: It's not obvious to me that we want to assert that no warnings (including ones from dependencies) were generated at any point before this test. If we want to ensure that no warnings (except the ones we expect) to be raised then I think we might want to explicitly add that to the pre/post test hooks?

Though I think it's a fair idea to try and figure out what the warning is, so I've pushed a commit that will fail if there were any existing warnings.

Comment (adiroiban) [reply src/twisted/test/test_twistd.py]: Yes... a post test hook would be great... but this needs to be done in a separate ticket/PR.

I have created this ticket a long time ago https://github.com/twisted/twisted/issues/10332 ... but didn't had time to look into it.

File: src/twisted/test/test_twistd.py [hunk 3]
```diff
@@ -1014,6 +1013,11 @@ def test_setUidSameAsCurrentUid(self):
         If the specified UID is the same as the current UID of the process,
         then a warning is displayed.
         """
+
+        # Assert that there were no existing warnings.
+        existing_warnings = self.flushWarnings()
+        self.assertEqual([], existing_warnings)
+
```
Comment (adiroiban) [inline src/twisted/test/test_twistd.py:1021]: Maybe add a link to the issue so that we will review this test once the test suite has some hooks for warnings.

```suggestion
        # FIXME:https://github.com/twisted/twisted/issues/10332
        # Assert that there were no existing warnings.
        existing_warnings = self.flushWarnings()
        self.assertEqual([], existing_warnings)

```

**Comment by adiroiban** [comment]
I have accidentally found this ticket https://github.com/twisted/twisted/issues/6635

not sure if it's related... or if this will fix it.

**Comment by erikjohnston** [comment]
Is `twisted.python.test.test_sendmsg.SendmsgTests.test_sendSubProcessFD` a known flakey test? I'm failing to a) reproduce it, and b) see how the changes would affect it.

```
Error: 
Traceback (most recent call last):
Failure: twisted.internet.defer.TimeoutError: <twisted.python.test.test_sendmsg.SendmsgTests testMethod=test_sendSubProcessFD> (test_sendSubProcessFD) still running at 120.0 secs

twisted.python.test.test_sendmsg.SendmsgTests.test_sendSubProcessFD
```

**Comment by glyph** [comment]
> Is `twisted.python.test.test_sendmsg.SendmsgTests.test_sendSubProcessFD` a known flakey test?

Yes, although I think we already tried to address this? https://github.com/twisted/twisted/issues/11830

I've re-run the tests for you, because yes, this seems unrelated.

File: src/twisted/test/test_defer.py [hunk 4]
```diff
@@ -1735,6 +1736,37 @@ def test_fromCoroutineRequiresCoroutine(self) -> None:
         for thing in thingsThatAreNotCoroutines:
             self.assertRaises(defer.NotACoroutineError, Deferred.fromCoroutine, thing)
 
+    def test_inlineCallbacksCancelCaptured(self) -> None:
+        """
+        Cancelling an `inlineCallbacks` correctly handles the function catching
```
Comment (adiroiban) [inline src/twisted/test/test_defer.py]: what does "correctly" mean?

I guess that we are writing any test to make sure the code "works correctly".

Can you please add a few more details about what is expected to happen in this test.

What does correctly means?

If this test would fail in the future, how can you tell which behaviour is correct and which is not?

File: src/twisted/test/test_defer.py [hunk 5]
```diff
@@ -1823,6 +1855,66 @@ def test_canceller_circular_reference_non_final(self) -> None:
         # been freed.
         self.assertIsNone(weakCanceller())
 
+    @pyunit.skipIf(_PYPY, "GC works differently on PyPy.")
+    def test_inlineCallbacksNoCircularReference(self) -> None:
+        """
+        Tests that there is no circular dependency when using
+        L{defer.inlineCallbacks}, so that the machinery gets cleaned up
+        immediately rather than waiting for a GC.
+        """
```
Comment (adiroiban) [inline src/twisted/test/test_defer.py]: Maybe something like this, focusing on the behaviour that is tested here.

We don't have any code in the test that looks into the GC part.

```suggestion
        """
        When using L{defer.inlineCallbacks},
        after the function exits,
        it will not keep references to the function itself or the arguments.
        """
```

File: src/twisted/test/test_defer.py [hunk 6]
```diff
@@ -1823,6 +1855,66 @@ def test_canceller_circular_reference_non_final(self) -> None:
...
+        # Create an object and weak reference to track if its gotten freed.
+        obj: Set[Any] = set()
+        objWeakRef = weakref.ref(obj)
+
+        async def func(a: Any) -> Any:
+            return a
+
+        # Run the function
+        funcD = Deferred.fromCoroutine(func(obj))
+        assert funcD.called
```
Comment (adiroiban) [inline src/twisted/test/test_defer.py]: Maybe this assertion is better.

It makes sure that the code returns the expected result... not only that it is called.

It can be called with a different result... or called with  failure.

```suggestion
        self.assertEqual(obj, self.successResultOf(funcD))
```

File: src/twisted/test/test_defer.py [hunk 7]
```diff
@@ -1823,6 +1855,66 @@ def test_canceller_circular_reference_non_final(self) -> None:
         # been freed.
         self.assertIsNone(weakCanceller())
 
+    @pyunit.skipIf(_PYPY, "GC works differently on PyPy.")
```
Comment (adiroiban) [inline src/twisted/test/test_defer.py]: can you add a comment about why this tests fails on PyPy?
I guess that it's expected for PyPy to work differently.
Maybe maybe something like:

`del instance` will not immediately free the object in PyPy.

-------

Maybe for people using PyPy this is common sense.
I have never used PyPy, so feel free to ignore this comment

File: src/twisted/internet/defer.py [hunk 8]
```diff
@@ -2085,6 +2105,48 @@ def gotResult(r: object) -> None:
             waiting[1] = None
 
 
+def _handleCancelInlineCallbacks(
```
Comment (adiroiban) [inline src/twisted/internet/defer.py:2124]: minor comment. just for readability.
Maybe move this after `def _addCancelCallbackToDeferred` as this looks like a help that is used only there

can it be nested in _addCancelCallbackToDeferred ? or this will not help with the circular reference fix?

File: src/twisted/test/test_defer.py [hunk 9]
```diff
@@ -1735,6 +1736,37 @@ def test_fromCoroutineRequiresCoroutine(self) -> None:
...
+
+        funcD = testFunc()
+        self.assertFalse(funcD.called)
+
+        funcD.cancel()
+
+        self.assertEqual(canceller1Calls, [d1])
+
+        funcD.cancel()
+        self.failureResultOf(funcD)
```
Comment (adiroiban) [inline src/twisted/test/test_defer.py]: Maybe it would help to also check which failure we have.

Maybe there is a bug and we get a TypeError or ValueError and I don't this this is what we are expecting.
```suggestion
        failure = self.failureResultOf(funcD)
        # Add a check for the type of failure.
```

File: src/twisted/test/test_defer.py [hunk 10]
```diff
@@ -1735,6 +1736,37 @@ def test_fromCoroutineRequiresCoroutine(self) -> None:
...
+                yield d1
+            except Exception:
+                pass
+
+            yield d2
+
+        funcD = testFunc()
+        self.assertFalse(funcD.called)
+
+        funcD.cancel()
```
Comment (adiroiban) [inline src/twisted/test/test_defer.py]: Maybe add a comment explaining the correct behaviour
```suggestion
        # On the first cancel it just handled by the
        # function by ignoring the exception.
        # The function still has no results as the
        # second deferred is not yet called.
```

File: src/twisted/test/test_defer.py [hunk 11]
```diff
@@ -1735,6 +1736,37 @@ def test_fromCoroutineRequiresCoroutine(self) -> None:
...
+
+        self.assertEqual(canceller1Calls, [d1])
```
Comment (adiroiban) [inline src/twisted/test/test_defer.py]: Maybe also add a check for canceller2Calls ?
```suggestion
        self.assertEqual(canceller1Calls, [d1])
        self.assertEqual(canceller2Calls, [])
        # Still waiting for the second deferred.
        self.assertNoResult(funcD)
```

File: src/twisted/test/test_defer.py [hunk 12]
```diff
@@ -1735,6 +1736,37 @@ def test_fromCoroutineRequiresCoroutine(self) -> None:
...
+
+        funcD.cancel()
```
Comment (adiroiban) [inline src/twisted/test/test_defer.py]:
```suggestion
        # Canceling again, will trigger the
        # cancellation errors as the second deferred
        # has no handling for cancellation.
        funcD.cancel()
```

**Review by adiroiban — CHANGES_REQUESTED** [review]
Many thanks for the cleanup. It looks good.

I only have a minor comment on the test part as I find it hard to understand what is going on there on the first read.

The code over here is a bit complex, so I think it's important to have good test to explain what is the expected behaviour...so that in the future it's easier to decide which behaviour is a bug and which is expected

Kind regards

File: src/twisted/internet/defer.py (hunk 8 above)
Comment (erikjohnston) [reply src/twisted/internet/defer.py:2124]: I have moved this function down.

> can it be nested in _addCancelCallbackToDeferred ? or this will not help with the circular reference fix?

Generally having nested functions can cause circular references, so I'm loathed to nest them. That said, nesting them doesn't cause the tests to fail, but that is probably because the objects caught by any circular reference here is limited to a small set of object we can't test for.

File: src/twisted/test/test_defer.py (hunk 4 above)
Comment (erikjohnston) [reply src/twisted/test/test_defer.py]: Good point. I've tried to clarify, let me know if it needs some correction.

File: src/twisted/test/test_defer.py (hunk 7 above)
Comment (erikjohnston) [reply src/twisted/test/test_defer.py]: I've added a note to the docstring.

**Comment by erikjohnston** [comment]
@adiroiban Thank you for the review! Hopefully the tests are a bit clearer now, but let me know if they're still a bit confusing as I'm happy to continue fixing them up (it is important).

And as always, thanks everyone for maintaining Twisted :slightly_smiling_face:

File: src/twisted/test/test_defer.py [hunk 13]
```diff
@@ -1735,6 +1736,61 @@ def test_fromCoroutineRequiresCoroutine(self) -> None:
...
+            3. Cancelling the deferred again cancels any deferred the function
+               is waiting on, and the exception is raised.
+        """
+        canceller1Calls: List[Deferred[Any]] = []
+        canceller2Calls: List[Deferred[Any]] = []
+        d1: Deferred[Any] = Deferred(canceller1Calls.append)
+        d2: Deferred[Any] = Deferred(canceller2Calls.append)
+
+        @defer.inlineCallbacks
+        def testFunc() -> Any:
```
Comment (graingert) [inline src/twisted/test/test_defer.py]: does this work with Deferred[object] isntead of Deferred[Any]?

```suggestion
        canceller1Calls: List[Deferred[object]] = []
        canceller2Calls: List[Deferred[object]] = []
        d1: Deferred[object] = Deferred(canceller1Calls.append)
        d2: Deferred[object] = Deferred(canceller2Calls.append)

        @defer.inlineCallbacks
        def testFunc() -> object:
```

Comment (erikjohnston) [reply src/twisted/test/test_defer.py]: Good point, yes it does. Albeit with a more complicated return type for `testFunc`

File: src/twisted/newsfragments/11885.misc [hunk 14]
```diff
@@ -0,0 +1 @@
+Fix circular references when using `defer.inlineCallbacks`.
```
Comment (adiroiban) [inline src/twisted/newsfragments/11885.misc]: Maybe we can call it a feature for CPython ... and name this file 11885.feature

```suggestion
`twisted.internet.defer.inlineCallbacks`, when using CPython`, after the deferred is called, it no longer keeps a reference to the wrapped function and arguments.
```

File: src/twisted/test/test_defer.py [hunk 15]
```diff
@@ -1735,6 +1736,61 @@ def test_fromCoroutineRequiresCoroutine(self) -> None:
         for thing in thingsThatAreNotCoroutines:
             self.assertRaises(defer.NotACoroutineError, Deferred.fromCoroutine, thing)
 
+    def test_inlineCallbacksCancelCaptured(self) -> None:
```
Comment (adiroiban) [inline src/twisted/test/test_defer.py]: Just asking

Is this test related to circular references?

Was this test created as there was regression that you discovered while refactoring the code, and the bug was not detected by the previous set of tests?

--------

I know that the existing code is already not ideal.

Maybe, as part of this PR, we can create a new `InlineCallbacksTests` class that will group all the tests targeting `inlineCallbacsk` usage

The current `DeferredTests` is a mess, as there are so many tests here and is hard to check for duplicate tests.

**Review by adiroiban — APPROVED** [review]
I think that it looks good.

The only minor comment is moving the tests targeting `inlineCallbacks` into a separate class, to make it easier to check the scope and coverage of the tests... and detect if we write duplicate tests for the same functionality/behaviour

File: src/twisted/test/test_defer.py (hunk 15 above)
Comment (erikjohnston) [reply src/twisted/test/test_defer.py]: The test isn't related to circular references, nor really a regression test. It's a new test that covers this particular edge case of cancellations that weren't covered by any other tests. It's a fiddly bit of logic, so didn't want to refactor it without more coverage.

---

Happy to move the tests into their own class. I'll do that now.

Comment (erikjohnston) [reply src/twisted/test/test_defer.py]: Done.

File: src/twisted/newsfragments/11885.misc (hunk 14 above)
Comment (erikjohnston) [reply src/twisted/newsfragments/11885.misc]: I've changed it to be a feature. I've tried to word it so that library users can better understand the impact, but I'm not sure if it's just too wordy. 

I'm happy for you to just change it if you want to. Words aren't my forte.

**Comment by erikjohnston** [comment]
There's a test failure on

> twisted.test.test_twistd.UnixApplicationRunnerStartApplicationTests.test_setUidSameAsCurrentUid

I think that must be unrelated?

**Comment by glyph** [comment]
Given that another test has been stuck in a pre-start state for 4 hours, I'm guessing that there was some blip in the CI runners.  Nothing on githubstatus.com, but, it could have been just on one host…

**Comment by glyph** [comment]
Kicking off a failed re-run and hopefully we can fix those flakes.

**Comment by erikjohnston** [comment]
Thanks!

> Given that another test has been stuck in a pre-start state for 4 hours, I'm guessing that there was some blip in the CI runners. Nothing on githubstatus.com, but, it could have been just on one host…

We've been seeing Github having issues all over the shop over the past few hours.
